pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeFilterOperator.java

```diff
@@ -0,0 +1,114 @@
+ */
+package org.apache.pinot.core.operator.filter;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.pinot.core.common.BlockDocIdIterator;
+import org.apache.pinot.core.common.DataSource;
+import org.apache.pinot.core.common.Predicate;
+import org.apache.pinot.core.operator.blocks.FilterBlock;
+import org.apache.pinot.core.operator.dociditerators.MVScanDocIdIterator;
+import org.apache.pinot.core.operator.dociditerators.SVScanDocIdIterator;
+import org.apache.pinot.core.operator.docidsets.BitmapDocIdSet;
+import org.apache.pinot.core.operator.docidsets.FilterBlockDocIdSet;
+import org.apache.pinot.core.operator.filter.predicate.PredicateEvaluator;
+import org.apache.pinot.core.operator.filter.predicate.RangePredicateEvaluatorFactory;
+import org.apache.pinot.core.segment.index.readers.RangeIndexReader;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+public class RangeFilterOperator extends BaseFilterOperator {
+  private static final String OPERATOR_NAME = "RangeFilterOperator";
+
+  private final PredicateEvaluator _predicateEvaluator;
+  private final DataSource _dataSource;
+  private final int _startDocId;
+  private final int _endDocId;
+  private final boolean _exclusive;
+
+  public RangeFilterOperator(PredicateEvaluator predicateEvaluator, DataSource dataSource, int startDocId,
+      int endDocId) {
+    // NOTE:
+    // Predicate that is always evaluated as true or false should not be passed into the BitmapBasedFilterOperator for
+    // performance concern.
+    // If predicate is always evaluated as true, use MatchAllFilterOperator; if predicate is always evaluated as false,
+    // use EmptyFilterOperator.
+    Preconditions.checkArgument(!predicateEvaluator.isAlwaysTrue() && !predicateEvaluator.isAlwaysFalse());
+
+    _predicateEvaluator = predicateEvaluator;
+    _dataSource = dataSource;
+    _startDocId = startDocId;
+    _endDocId = endDocId;
+    _exclusive = predicateEvaluator.isExclusive();
+  }
+
+  @Override
+  protected FilterBlock getNextBlock() {
+
+    //only dictionary based is supported for now
+    Preconditions.checkState(_predicateEvaluator instanceof RangePredicateEvaluatorFactory.OfflineDictionaryBasedRangePredicateEvaluator);
+
+    RangePredicateEvaluatorFactory.OfflineDictionaryBasedRangePredicateEvaluator evaluator =
+        (RangePredicateEvaluatorFactory.OfflineDictionaryBasedRangePredicateEvaluator) _predicateEvaluator;
+
+    RangeIndexReader rangeIndexReader = (RangeIndexReader) _dataSource.getRangeIndex();
+    int startRangeId = rangeIndexReader.findRangeId(evaluator.getStartDictId());
+    int endRangeId = rangeIndexReader.findRangeId(evaluator.getEndDictId());
+    //Handle Matching Ranges - some ranges match fully but some partially
+    //below code assumes first and last range always match partially which may not be the case always //todo: optimize it
```

Comment:
This is not about optimization. Won't this impact correctness?
For a given startDictId and endDictId, we find the matching ranges represented by startRangeId and endRangeId. 
Some/all ranges between these could be a full match and some/all could be a partial match. Right?
I understand that all full matches can be ORed but how can we assume that everything from startRangeId + 1 to endRangeId - 1 will be a full match?

Reply:
why not?